### ISSUE TITLE
OpenTK 3.x crashes MacOS Monterey Beta

### ISSUE THREAD

**Issue by BleuBleu** [opentk/opentk#1308]
### Description

Hi. First of all, **huge apologies** for both opening an issue for an old version (3.x) AND for an OS that is in Beta :-)  

Please consider this more of a brainstorm to find a solution. I'm in a situation where i just cannot upgrade to 4.x, so i need to find a solution, even if it means recompiling myself a custom version of 3.x. 

Any ideas will be appreciated.

**EDIT** : After a bit of debugging, it seems like on Monterey, NSLookupSymbolInImage() always return NULL, even though NSAddImage worked properly. dlsym seems to still work tho. Is there a MacOS expert that could chime in and explain why that would be the case?

### Repro steps

1. Create a version simple project with a game window with OpenTK 3.x (Tried 3.1.0 and 3.3.1), run it on MacOS Monterey Beta 1 or 2, something like this will do:

```
    class TestWindow : GameWindow
    {
        protected override void OnRenderFrame(FrameEventArgs e)
        {
            base.OnRenderFrame(e);

            GL.Disable(EnableCap.DepthTest);
            GL.Viewport(0, 0, Width, Height);
            GL.ClearColor(1, 0, 1, 1);
            GL.Clear(ClearBufferMask.ColorBufferBit);

            SwapBuffers();
        }
    }
```

2. Create + Run() that window
3. Crash.

### Expected behavior

Ive been using OpenTK for a while on MacOS and it worked fined with Catalina, Big Sur and some older versions. Super solid and stable. Not sure what Apple may have changed in Monterey that breaks everything. 

In understand that OpenGL is deprecated on MacOS, but it is still working, although will receive zero improvements in the future and might disappear one day. But as of right now, it still *works*.

Basically, the OpenGL context seems to create fine, but the very first real GL call you do will lead to a crash.

### Actual behavior

The window appear (with a gray background), then we crash on the first GL call. 

The mono log is totally not helpful. I wonder if we are dealing with some sort of stomp.
[log.txt](https://github.com/opentk/opentk/files/6766896/log.txt)

### Related information 

* Operating system : MacOS Monterey Beta 1/2
* SDL or Native backend : Cocoa/GL/mono 6.12
* What version of OpenTK (Installer, NuGet, Source Code) : 3.1.0 and 3.3.1 tested.
* What runtime (.NET Framework, CoreCLR, Mono) : .NET Framework 4.6.2
* Workarounds (if any) : None

**Comment by NogginBops** [comment]
OpenTK 3 is still considered "supported" and we will merge fixes for it and make new releases with said fixes, but we (the maintainers) will likely not spend significant time trying to fix the issue.

It doesn't seem like you've included what kind of error message you get with the crash, or does it crash silently with no output in the console?

I'm not a macOS expert so I don't know, but my guess is that this is a mono issue related to dllmap. As OpenTK is not doing any calls to the macOS api directly.

**Comment by NogginBops** [comment]
It would also be helpful if you could include information on how to got to the conclusion that `NSLookupSymbolInImage()` returns `NULL`.

**Comment by BleuBleu** [comment]
> It doesn't seem like you've included what kind of error message you get with the crash, or does it crash silently with no output in the console?

I included log.txt which is what mono outputs. Its not helpful at all. 

> It would also be helpful if you could include information on how to got to the conclusion that `NSLookupSymbolInImage()` returns `NULL`.

I compiled that latest 3.x branch myself, added a ungodly amount of logging and ran it on both Big Sur (current version) and Monterey Beta. I saw that Monterey was getting NULL on all symbols. I then proceeded to replace the NSLookupSymbolInImage by a good-old dlym and that fixed the issue. Got OpenTK to run again!

**Here is what i propose**: Let's wait until the real version of Monterey comes out, maybe Apple will fix this. Not worth doing anything drastic for a Beta.
* If they do, then its all good.
* If they dont, ill submit a PR to fallbackto dlsym when NSLookupSymbolInImage fails.

Sounds good?

-Mat

**Comment by NogginBops** [comment]
I didn't know OpenTK 3 used `NSLookupSymbolInImage` when getting OpenGL procs.
I looked up some documentation on these functions that dated back to 2014, which has this to say.
![image](https://user-images.githubusercontent.com/3778357/124701482-f80b3980-deee-11eb-8da8-772c5a977255.png)
And the referenced section contains this 
![image](https://user-images.githubusercontent.com/3778357/124701514-048f9200-deef-11eb-8a8b-181b92a06f63.png)
So I don't think we can conclude that `dlopen` and `dlsym` are the "correct" functions that should have been used, I'm guessing this code is probably very old (tao framework 2009 old).

While looking around I found this thread discussing how to do the changes
https://sourceforge.net/p/glew/patches/17/
it suggest that you should also remove the code that perpends an underscore to the name before sending it to `dlsym`.

It would be great if you could make a PR with this change. :)
If you have any more MacOS related stuff we would love to have your assistance, as the MacOS backend in OpenTK 3 doesn't receive a lot of love, and there are probably a few bugs left to iron out.

**Comment by BleuBleu** [comment]
Wow thanks for the extensive research!!! Amazing!

Yes, when i tried with dlsym, i also noticed it didn't need the underscore prefix, so that's good to have this confirmed.

There is one comment in the code mentioning NSLookupSymbolInImage is faster than dlsym, ill see if I can get some numbers of that and report back once I put something together. 

```
        // Unfortunately, these are slower even if they are more
        // portable and simpler to use.
        [DllImport(Library)]
        internal static extern IntPtr dlopen(String fileName, int flags);
        ...
```

I probably only use about 1% of the features of OpenTK, so i cant say i have many other improvements to suggest for MacOS.

In the short term, ill focus on this issue!

-Mat

**Comment by BleuBleu** [comment]
Ok, I have the change ready, but there is a bit of a problem and I'd like your input.

The performance is much worse, as the comment alluded to. The time of LoadAll(), which processes all the bindings, goes from 40ms to 900ms on my old 2014 Mac Mini. If it was 2-3x more expensive, i wouldnt care much but we are talking about like 20x more! :) I tried various things, different flags to dlopen, etc. Cost stays the same.

I can think of a few solution:
- Not care about it :)
- Give the programmer the option, some kind of "BindingMethod" that would be dlsym or NSLookup, although, this seem error prone
- Try to auto-detect if NSLookup works on the current OS by looking up a very common entry point at startup and see if works (like glClear or something), then fallbacking to dlsym if it doesnt. 

Any suggestions are appreciated. I spent more time playing with NSLookup on Monterey Beta, and it just doesnt work.

-Mat

**Comment by NogginBops** [comment]
Does the slowdown apply to Monterey too? Or is it only for the old version of MacOS where it's slow?

**Comment by BleuBleu** [comment]
> Does the slowdown apply to Monterey too? Or is it only for the old version of MacOS where it's slow?

MacOS Big Sur running on a Mac Mini 2014
- NSLookup : 40ms
- dlsym : 900ms

MacOS Monterey inside a VM on a 3950x
- NSLookup : 15ms (Does not work, always return NULL)
- dlsym : 650ms

**Comment by NogginBops** [comment]
oof, that is a considerable slowdown. Are there any other ways to load dylib functions in MacOS? I can't imagine that 650ms is an acceptable amount of time for an operating system to do a few (but not too many) simple symbol lookups...

**Comment by BleuBleu** [comment]
This is my first project where i have to support MacOS so your guess is as good as mine. Looking a the doc very quickly these seem to be the main 2 methods. Technically there are also the _dyld_xxx functions, but these are what dlsym is built on top of, i doubt there is a significant amount of overhead we can shave by going lower-level.

-Mat

**Comment by NogginBops** [comment]
I will take a look at this more closely tomorrow. I can't really test it as I don't have a mac to test on, but maybe I can find something intersting.

**Comment by BleuBleu** [comment]
Cool. Let me know if you want me to try stuff.

-Mat

**Comment by NogginBops** [comment]
I couldn't really find any better alternative for loading symbols on MacOS.
It looks like Apple want people to use weak references and other stuff that is hard to achieve easily when using mono and other similar cross platform JIT runtimes (you need to emit special symbol attributes ` __attribute__((weak_import))` for it to work).

I think this workaround that you posted above where you try figure out what function to use for symbol resolution is a good workaround for this case. It's an unfortunate reality, but we don't want to add ~1s startup time for MacOS users if we don't need to, especially as there are quite a few projects using OpenTK 3.

**Comment by NogginBops** [comment]
Also, pinging @tyronx as it might be interesting to know that this might be a future issue in your code. 🙂

**Comment by BleuBleu** [comment]
Got it. 

Ill prepare the PR tonight or over the weekend. 

Keep in mind also, a 2014 mac mini is pretty much one of the worst Mac that can still run stuff :) I wouldn't be surprised the average macbook runs twice as fast. Altough 500ms is still horrible just to load function pointers. :)

I also spent some time looking to see if there was a function that could simply list all of the symbols + pointers, and couldnt find one. I think on Linux there is "dlinfo" but of course, if doesn't exist on MacOS. We could have parsed the whole list at startup and cache it, that would have been very nice.

-Mat

**Comment by NogginBops** [comment]
Yeah, it's loading 1360 symbols which realistically should take around 0.1 ms (with a good hash map, excluding the time to load the map from disk, but that should also be minimal) but it does seem like MacOS has some ***really*** inefficient lookup code.

It would be great if you could make a PR!

**Comment by BleuBleu** [comment]
Got it. 

Ill prepare the PR tonight or over the weekend. 

Keep in mind also, a 2014 mac mini is pretty much one of the worst Mac that can still run stuff :) I wouldn't be surprised the average macbook runs twice as fast. Altough 500ms is still horrible just to load function pointers. :)

I also spent some time looking to see if there was a function that could simply list all of the symbols + pointers, and couldnt find one. I think on Linux there is "dlinfo" but of course, if doesn't exist on MacOS. We could have parsed the whole list at startup and cache it, that would have been very nice.

-Mat

**Comment by UnknownShadow200** [comment]
Does using `RTLD_FIRST` for the flags passed to `dlopen` make any difference?

**Comment by BleuBleu** [comment]
> Does using `RTLD_FIRST` for the flags passed to `dlopen` make any difference?

Woah!!! I missed that flag in my research.

Time for LoadAll(), on a Mac Mini 2014:

- NSLookup: 40ms
- dysym (with RTLD_NOW) : 900ms
- dlsym (with RTLD_NOW | RTLD_FIRST) : 55ms

I confirm it still fixes our issues with the Monterey beta.

This is a game changer! If you ask me we should just ditch the whole "autodetection" scheme and simply switch to dlsym with this flag. 

What do you guys think?

-Mat

**Comment by NogginBops** [comment]
Using `dlopen` with `RLTD_FIRST` does sound like the preferred solution now as it doesn't increase load times too significantly (15ms is quite a lot but fine for this I think). 
So I think we should go for this solution as it makes the code work and doesn't slow it down too much.

**Comment by BleuBleu** [comment]
> So I think we should go for this solution as it makes the code work and doesn't slow it down too much.

Will update the PR and re-test everything.

-Mat

**Comment by BleuBleu** [comment]
Done. 

Also, as mentioned in the code, I took the liberty of opening a "Feedback Assistant" issue (that's how you report beta bugs) with Apple. FB9319518. I don't expect anything, but we shall see.

-Mat

**Comment by BleuBleu** [comment]
Hi! I wanted to check with you what kind of timeline you had for 3.3.2. I'm planning a beta of my app in a couple weeks and id be nice to have that fix it. If you guys need more time for 3.3.2 no worries at all, i can ship with a custom built OpenTK 3.3.1 until we have an official 3.3.2. 

No pressure really! :)

-Mat

**Comment by BleuBleu** [comment]
Quick update: Received response from Apple! We did the right thing apparently.

![image](https://user-images.githubusercontent.com/6654334/126261667-f8db8aaf-3ccc-4e99-a5f7-ed84963fd3a2.png)
